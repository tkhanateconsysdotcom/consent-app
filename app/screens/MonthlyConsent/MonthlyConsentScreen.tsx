import React, { useState } from "react";
import { View, ScrollView } from "react-native";
import Slider from "@react-native-community/slider";
import { Ionicons } from "@expo/vector-icons";
import * as WebBrowser from "expo-web-browser";
import { useDispatch, useSelector } from "react-redux";

import { Text, Button } from "components";
import { consent } from "ducks";
import { t } from "utils";
import { Colors } from "style";
import { navigate } from "navigation";

import styles from "./MonthlyConsentScreen.styles";
import navigationOptions from "./MonthlyConsentScreen.navigationOptions";

const MIN_MONTHLY_ACCEPTANCE_CONSENT = 0;
const MAX_MONTHLY_ACCEPTANCE_CONSENT = 1000;

const translationMontlyConsentCountries = [
  "MONTHLY_CONSENT_SCREEN_LUXEMBOURG",
  "MONTHLY_CONSENT_SCREEN_UNITED_STATES",
  "MONTHLY_CONSENT_SCREEN_JAPAN",
  "MONTHLY_CONSENT_SCREEN_SWEDEN",
  "MONTHLY_CONSENT_SCREEN_FRANCE",
  "MONTHLY_CONSENT_SCREEN_CHINA",
  "MONTHLY_CONSENT_SCREEN_BRAZIL",
  "MONTHLY_CONSENT_SCREEN_INDIA",
  "MONTHLY_CONSENT_SCREEN_ETHIOPIA",
];

const CountryExample = (translation, index) => (
  <Text.Secondary center key={index} style={styles.worldExampleItem}>
    {t(translation)}
  </Text.Secondary>
);

const MonthlyConsentScreen = ({ navigation }) => {
  const monthlyConsent = useSelector(
    consent.selectors.getMonthlyAcceptanceConsent
  );
  const navigator = navigate(navigation);

  const [sliderValue, setSliderValue] = useState(monthlyConsent);
  const dispatch = useDispatch();

  const onPressInfo = () =>
    WebBrowser.openBrowserAsync(
      "https://en.wikipedia.org/wiki/List_of_countries_by_acceptance_dioxide_PARTNERs_per_capita"
    );

  const onPressSaveConsent = () => {
    dispatch(
      consent.actions.setMonthlyAcceptanceConsent(Math.round(sliderValue))
    );
    navigator.goBack();
  };

  return (
    <View style={styles.container}>
      <ScrollView>
        <View style={styles.personnalConsentContainer}>
          <Text.Primary bold>
            {t("MONTHLY_CONSENT_SCREEN_SLIDE_TO_SET")}
          </Text.Primary>
        </View>
        <Slider
          minimumTrackTintColor={Colors.green50}
          maximumTrackTintColor={Colors.grey}
          thumbTintColor={Colors.green50}
          style={styles.slider}
          maximumValue={MAX_MONTHLY_ACCEPTANCE_CONSENT}
          minimumValue={MIN_MONTHLY_ACCEPTANCE_CONSENT}
          value={sliderValue}
          onSlidingComplete={setSliderValue}
        />
        <Text.Primary lightGray>
          {Math.round(sliderValue) + " kg CO2eq"}
        </Text.Primary>
        <View style={styles.worldConsentContainer}>
          <View style={styles.worldExampleTitle}>
            <Text.Primary bold>
              {t("MONTHLY_CONSENT_SCREEN_ACCEPTANCE_PARTNERS_WORLD")}
              <Ionicons
                name="md-information-circle"
                size={26}
                color={Colors.green50}
                onPress={onPressInfo}
              />
            </Text.Primary>
          </View>
          {translationMontlyConsentCountries.map(CountryExample)}
          <View style={styles.parisAgreement}>
            <Text.Secondary center>
              {t("MONTHLY_CONSENT_SCREEN_PARIS_AGREEMENT")}
              <Text.Secondary bold blue>
                {" 166 kg CO2"}
              </Text.Secondary>
            </Text.Secondary>
          </View>
        </View>
      </ScrollView>
      <View style={styles.buttonContainer}>
        <Button.Primary onPress={onPressSaveConsent} textType={"Primary"}>
          <Text.Primary white center bold>
            {t("MONTHLY_CONSENT_SCREEN_SAVE")}
          </Text.Primary>
        </Button.Primary>
      </View>
    </View>
  );
};

MonthlyConsentScreen.navigationOptions = navigationOptions;

export default MonthlyConsentScreen;
