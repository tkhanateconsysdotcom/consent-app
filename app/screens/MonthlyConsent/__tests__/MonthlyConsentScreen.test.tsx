import React from "react";
import { create } from "react-test-renderer";

import MonthlyConsentScreen from "../MonthlyConsentScreen";

const props = {
  navigation: {
    goBack: jest.fn(),
  },
};

it("MonthlyConsentScreen renders correctly", () => {
  const tree = create(<MonthlyConsentScreen {...props} />).toJSON();
  expect(tree).toMatchSnapshot();
});

it("MonthlyConsentScreen should go back if save button is pressed", () => {
  const root = create(<MonthlyConsentScreen {...props} />).root;
  // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
  // @ts-ignore
  const button = root.findByType("Button.Primary");

  button.props.onPress();
  expect(props.navigation.goBack).toHaveBeenCalled();
});
