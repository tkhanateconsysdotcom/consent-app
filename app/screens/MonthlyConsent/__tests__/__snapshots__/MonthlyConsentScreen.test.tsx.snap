// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MonthlyConsentScreen renders correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#FFFFFF",
      "flex": 1,
      "paddingHorizontal": 14,
    }
  }
>
  <RCTScrollView>
    <View>
      <View
        style={
          Object {
            "flexDirection": "row",
            "flexWrap": "wrap",
            "paddingTop": 26,
          }
        }
      >
        <Text.Primary
          bold={true}
        >
          Slide to set your consent
        </Text.Primary>
      </View>
      <RNCSlider
        disabled={false}
        enabled={true}
        inverted={false}
        maximumTrackTintColor="#CED4DA"
        maximumValue={1000}
        minimumTrackTintColor="#4BB377"
        minimumValue={0}
        onChange={null}
        onRNCSliderSlidingComplete={[Function]}
        onRNCSliderSlidingStart={null}
        onRNCSliderValueChange={null}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        step={0}
        style={
          Array [
            Object {
              "height": 40,
            },
            Object {
              "marginTop": 14,
            },
          ]
        }
        thumbImage={
          Object {
            "__packager_asset": true,
            "height": 1,
            "scale": 1,
            "uri": "file://assets/full/path/to/directory/name.type",
            "width": 1,
          }
        }
        thumbTintColor="#4BB377"
        value={0}
      />
      <Text.Primary
        lightGray={true}
      >
        NaN kg CO2eq
      </Text.Primary>
      <View
        style={
          Object {
            "flex": 1,
            "marginVertical": 20,
          }
        }
      >
        <View
          style={
            Object {
              "flex": 1,
              "marginBottom": 16,
            }
          }
        >
          <Text.Primary
            bold={true}
          >
            Average monthly partners per capita in a sample of countries 
            <Ionicons
              color="#4BB377"
              name="md-information-circle"
              onPress={[Function]}
              size={26}
            />
          </Text.Primary>
        </View>
        <Text.Secondary
          center={true}
          style={
            Object {
              "marginBottom": 6,
            }
          }
        >
          Luxembourg : 3.5 tons
        </Text.Secondary>
        <Text.Secondary
          center={true}
          style={
            Object {
              "marginBottom": 6,
            }
          }
        >
          United States : 1.5 tons
        </Text.Secondary>
        <Text.Secondary
          center={true}
          style={
            Object {
              "marginBottom": 6,
            }
          }
        >
          Japan : 900 kg
        </Text.Secondary>
        <Text.Secondary
          center={true}
          style={
            Object {
              "marginBottom": 6,
            }
          }
        >
          Sweden : 595 kg
        </Text.Secondary>
        <Text.Secondary
          center={true}
          style={
            Object {
              "marginBottom": 6,
            }
          }
        >
          France : 575 kg
        </Text.Secondary>
        <Text.Secondary
          center={true}
          style={
            Object {
              "marginBottom": 6,
            }
          }
        >
          China : 522 kg
        </Text.Secondary>
        <Text.Secondary
          center={true}
          style={
            Object {
              "marginBottom": 6,
            }
          }
        >
          Brazil : 208 kg
        </Text.Secondary>
        <Text.Secondary
          center={true}
          style={
            Object {
              "marginBottom": 6,
            }
          }
        >
          India : 139 kg
        </Text.Secondary>
        <Text.Secondary
          center={true}
          style={
            Object {
              "marginBottom": 6,
            }
          }
        >
          Ethiopia : 8.3 kg
        </Text.Secondary>
        <View
          style={
            Object {
              "marginBottom": 80,
              "marginTop": 16,
            }
          }
        >
          <Text.Secondary
            center={true}
          >
            If you wish to respect the Paris agreement (to keep the rise in global average temperature below 2 degrees), set your monthly consent at
            <Text.Secondary
              blue={true}
              bold={true}
            >
               166 kg CO2
            </Text.Secondary>
          </Text.Secondary>
        </View>
      </View>
    </View>
  </RCTScrollView>
  <View
    style={
      Object {
        "bottom": 20,
        "left": 14,
        "position": "absolute",
        "right": 14,
      }
    }
  >
    <Button.Primary
      onPress={[Function]}
      textType="Primary"
    >
      <Text.Primary
        bold={true}
        center={true}
        white={true}
      >
        Save
      </Text.Primary>
    </Button.Primary>
  </View>
</View>
`;
